File: examples/contrib/set_covering2.cs
//
// Copyright 2012 Hakan Kjellerstrand
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

using System;
using System.Collections;
using System.IO;
using System.Text.RegularExpressions;
using Google.OrTools.ConstraintSolver;

public class SetCovering2
{
    /**
     *
     * Solves a set covering problem.
     * See  See http://www.hakank.org/or-tools/set_covering2.py
     *
     */
    private static void Solve()
    {
        Solver solver = new Solver("SetCovering2");

        //
        // data
        //

        // Example 9.1-2 from
        // Taha "Operations Research - An Introduction",
        // page 354ff.
        // Minimize the number of security telephones in street
        // corners on a campus.

        int n = 8;            // maximum number of corners
        int num_streets = 11; // number of connected streets

        // corners of each street
        // Note: 1-based (handled below)
        int[,] corner = { { 1, 2 }, { 2, 3 }, { 4, 5 }, { 7, 8 }, { 6, 7 }, { 2, 6 },
                          { 1, 6 }, { 4, 7 }, { 2, 4 }, { 5, 8 }, { 3, 5 } };

        //
        // Decision variables
        //
        IntVar[] x = solver.MakeIntVarArray(n, 0, 1, "x");
        // number of telephones, to be minimized
        IntVar z = x.Sum().Var();

        //
        // Constraints
        //

        // ensure that all streets are covered
        for (int i = 0; i < num_streets; i++)
        {
            solver.Add(x[corner[i, 0] - 1] + x[corner[i, 1] - 1] >= 1);
        }

        //
        // objective
        //
        OptimizeVar objective = z.Minimize(1);

        //
        // Search
        //
        DecisionBuilder db = solver.MakePhase(x, Solver.INT_VAR_DEFAULT, Solver.INT_VALUE_DEFAULT);

        solver.NewSearch(db, objective);

        while (solver.NextSolution())
        {
            Console.WriteLine("z: {0}", z.Value());
            Console.Write("x: ");
            for (int i = 0; i < n; i++)
            {
                Console.Write(x[i].Value() + " ");
            }
            Console.WriteLine();
        }

        Console.WriteLine("\nSolutions: {0}", solver.Solutions());
        Console.WriteLine("WallTime: {0}ms", solver.WallTime());
        Console.WriteLine("Failures: {0}", solver.Failures());
        Console.WriteLine("Branches: {0} ", solver.Branches());

        solver.EndSearch();
    }

    public static void Main(String[] args)
    {
        Solve();
    }
}

File: examples/contrib/set_covering2.py
# Copyright 2010 Hakan Kjellerstrand hakank@gmail.com
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Set covering in Google CP Solver.

Example 9.1-2, page 354ff, from
Taha 'Operations Research - An Introduction'
Minimize the number of security telephones in street
corners on a campus.

Compare with the following models:
* MiniZinc: http://www.hakank.org/minizinc/set_covering2.mzn
* Comet   : http://www.hakank.org/comet/set_covering2.co
* ECLiPSe : http://www.hakank.org/eclipse/set_covering2.ecl
* SICStus: http://hakank.org/sicstus/set_covering2.pl
* Gecode: http://hakank.org/gecode/set_covering2.cpp

This model was created by Hakan Kjellerstrand (hakank@gmail.com)
Also see my other Google CP Solver models:
http://www.hakank.org/google_or_tools/
"""

from ortools.constraint_solver.python import constraint_solver as cp


def main(unused_argv):

    # Create the solver.
    solver = cp.Solver("Set covering")

    #
    # data
    #
    n = 8  # maximum number of corners
    num_streets = 11  # number of connected streets

    # corners of each street
    # Note: 1-based (handled below)
    corner = [
        [1, 2],
        [2, 3],
        [4, 5],
        [7, 8],
        [6, 7],
        [2, 6],
        [1, 6],
        [4, 7],
        [2, 4],
        [5, 8],
        [3, 5],
    ]

    #
    # declare variables
    #
    x = [solver.new_int_var(0, 1, "x[%i]" % i) for i in range(n)]

    #
    # constraints
    #

    # number of telephones, to be minimized
    z = solver.sum(x)

    # ensure that all corners are covered
    for i in range(num_streets):
        # also, convert to 0-based
        solver.add_sum_greater_or_equal([x[j - 1] for j in corner[i]], 1)

    objective = solver.minimize(z.var(), 1)

    #
    # solution and search
    #
    solution = solver.assignment()
    solution.add(x)
    solution.add_objective(z.var())

    collector = solver.last_solution_collector(solution)
    solver.solve(
        solver.phase(
            x,
            cp.IntVarStrategy.INT_VAR_DEFAULT,
            cp.IntValueStrategy.INT_VALUE_DEFAULT,
        ),
        [collector, objective],
    )

    print("z:", collector.objective_value(0))
    print("x:", [collector.value(0, x[i]) for i in range(n)])

    print("failures:", solver.num_failures)
    print("branches:", solver.num_branches)
    print("WallTime:", solver.wall_time_ms)


if __name__ == "__main__":
    main("cp sample")
